### DISCUSSION TITLE
RTLMP_FLOW=True and HIREARCHICAL_SYNTHESIS=1

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#4532] [General]
Why are these used in combination?

What happens if I don't?

https://github.com/The-OpenROAD-Project/OpenROAD/discussions/4529 has an fMax of 125, which isn't enormously impressive. There are 79 levels of logic in the longest path, which indicates to me that something isn't being optimized in synthesis...

BOOMv documents that it relies heavily on [register retiming](https://docs.boom-core.org/en/latest/sections/physical-realization.html#physical-realization) and I've also observed that it needs [register cloning](https://github.com/The-OpenROAD-Project/OpenROAD/discussions/4522#discussioncomment-8106429) for synchronous reset.

Hence, I want to try disabling hierarchical synthesis on MegaBoom, but given the long turnaround times, 40 hours, it would be nice to know a little bit about what I'm doing and what I should expect...

**Comment by rovinski** [comment]
I believe that RTLMP requires hierarchy in the netlist to work at all. If the netlist is flat, RTLMP won't work.

**Reply by maliberty** [reply]
no it will use tritonpart in that case but it will benefit from hier

**Reply by oharboe** [reply]
Can you elaborate a bit?

What is "tritonpart" compared to rtlmp?

If I disable hierarchical synthesis, I should expect worse results for macro placement, but better due to better synthesis for megaboom?

**Reply by maliberty** [reply]
TritonPart (par) is a hypergraph partitioner.  When rtlmp run into a large leaf block that it wants to break down and has no further rtl hierarchy to use then it uses traditional partitioning.  Having hierarchy may lead to different results and will generally be better.

The ideal is to flatten small modules to get good optimization while keeping large modules for both rtlmp and user inspection.  MAX_UNGROUP_SIZE helps with that tradeoff but note that it has to run yosys twice (one to get modules sizes, again with the selective flattening).

**Reply by oharboe** [reply]
I think I understand somewhat.

The reason I want to flatten is to get e.g. the synchronous reset to be laid out better, which means flattening large modules...  Also there are other concerns with large modules(forwarding of signals across pipeline stages). The policy of keeping modules based upon size doesn't have the sophistication to cover this case.

As to running twice, I was wondering, since this is a very simple policy anyway and the only thing that drives the policy is area, would a rough estimate of area be good 'nuf?

Looking at https://yosyshq.readthedocs.io/projects/yosys/en/latest/cmd/synth.html, running only the two first steps "hierarchy" + "proc" might be enough to get good enough area numbers to drive the policy.

```
    begin:
        hierarchy -check [-top <top> | -auto-top]

    coarse:
        proc
[stop here]
````

Some devil in detail, but perhaps this could be a fast way to get roughly the same quality of results...

**Reply by maliberty** [reply]
@gadfort has some ideas on getting it earlier but I haven't spent the time to see if they pan out across all cases.  If you want to investigate further that would be good.  A rough approximation is good enough but you do have to get far enough in the mapping to resolve high level operations.

**Reply by oharboe** [reply]
Silly question, what are "high level operations"?

**Reply by oharboe** [reply]
Let's see what [this](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1755) does... Didn't measure the running times, I just want some feedback from CI before spending more time on it.

**Reply by maliberty** [reply]
For example you need to map operators like add & mult to something much lower level to get an area estimate.

**Reply by oharboe** [reply]
There are no tests that can distinguish between the quality of results of what's in master and this PR: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1755

**Reply by maliberty** [reply]
At least no public tests.  We would have to run the private ones as well.

**Reply by oharboe** [reply]
Hold off on that until I have some numbers as to whether it is faster :-)

**Reply by oharboe** [reply]
The techmap stage is the one that takes the most time, it seems.

I wouldn't expect there to be any area estimates before techmap, but perhaps even before techmap there is some number(not really area) that is good 'nuf to drive the policy and yield the same quality of results?

**Reply by oharboe** [reply]
RTLMP fails on MegaBoom without hierarchical synthesis: https://github.com/The-OpenROAD-Project/OpenROAD/issues/4539

**Reply by oharboe** [reply]
The running times of https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1755 are almost the same as today, so this didn't pan out immediately. Without technology map, which takes a lot of time, there's no area estimate.

**Reply by oharboe** [reply]
@maliberty Tinkered a bit with it... https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1758

**Comment by oharboe** [comment]
> At least no public tests. We would have to run the private ones as well.

**Comment by gadfort** [comment]
@oharboe the retiming doesn't work in Yosys (and not in OpenROAD), so that is going to be a limitation. We ran into this as well with something else from the Chisel team folks and there isn't a good solution at the moment

**Reply by oharboe** [reply]
I see. I'm thinking, for now, retiming has to be done manually in the RTL (Chisel for MegaBoom).

What about register cloning, is that done in yosys?

My main effort currently is to get MegaBoom running nicely in a Bazel layer on top of ORFS, but I'm walking past all sorts of rabbitholes... Bazel offers artifacts and much more fine grained dependencies than ORFS.

**Reply by gadfort** [reply]
register cloning is something that is driven by physically aware constraints, so you really wouldn't want yosys to do that. For retiming that would require a more fundamental change to how yosys and abc interface to eachother and see if abc supports retiming

**Reply by oharboe** [reply]
I see. At least with flattening the registers in the synchronous reset aren't hidden inside the top module. I can see how register retiming and cloning would have to be more explicit in the RTL for now. Yes, it's abstraction leakage, but for now, it is what it is.

**Reply by maliberty** [reply]
fwiw abc does have a retime command

```
retime – Implements several flavors of retiming: most forward, most backward, a [minimum-register](http://www.eecs.berkeley.edu/~alanmi/publications/2007/fmcad07_min.pdf), a heuristic minimum-delay, the delay-optimal retiming presented in Peichen Pan, “Continuous retiming: algorithms and applications”, ICCD 1997, pp. 116-121. The latches are optimally shared across the fanout stems when the circuit is transformed from the sequential AIG into a logic network. The computation of initial states after retiming is reduced to a SAT problem, which is solved using [MiniSat](http://www.cs.chalmers.se/Cs/Research/FormalMethods/MiniSat/). The known issues include non-verifies after retiming with latches, which have don’t-care initial states. (For the time being, don’t-care initial states can be replaced by fixed initial states using command init.)
```
https://people.eecs.berkeley.edu/~alanmi/abc/

**Reply by gadfort** [reply]
@maliberty that is good to know. I was wondering about that when talking with @precisionmoon earlier
